### PR TITLE
feat: add `enarx deploy` subcommand

### PR DESCRIPTION
**Pull request by rvolosatovs**
Closes https://github.com/profianinc/drawbridge/issues/119

This PR introduces a concept of a package (which is an overloaded term, but we can address naming/branding later). A package is a tree-like structure, e.g. a directory, containing an `Enarx.toml` and `main.wasm`.
`enarx` can execute either a package, using the contained `Enarx.toml` within its root or a WASM binary directly.

That means:
```
enarx deploy file:///path/to/my.wasm
```
is a valid invocation, as is
```
enarx deploy file:///path/to/my/package
```
Where `/path/to/my/package` is an absolute path to a local directory containing (at least) `Enarx.toml` and `main.wasm`

Similarly, `http` and `https` schemes are supported, e.g.
```
enarx deploy https://example.com/path/to/my.wasm
```

If the response of a GET request to specified URL has a `Content-Type` of:
- `application/wasm` (need to double check the media type, if one exists already) => it attempts to execute it
- `application/vnd.drawbridge.directory.v1+json` => it assumes it's a https://github.com/profianinc/drawbridge tree directory and attempts to fetch `main.wasm` and `Enarx.toml` from the node rooted at specified URL.

Other media types, and schemes may be supported in the future.

In pure-WASM execution mode (i.e. not a package), the default config is implicitly used.

Apart from that, the PR adds a two-way communication channel between the exec layer and the host binary via unnamed Unix sockets. Potentially this could be used to send e.g. logs or errors to the host if desired. A TOML-encoded config is sent once on the socket by the host.

For uploading to Drawbridge, use:
```sh
#!/usr/bin/env bash
set -ex

# Create repository
curl --fail-with-body -sX PUT -H "Content-Type: application/json" -d '{}' "localhost:8080/${1}"

# Create tag
curl --fail-with-body -sX PUT -H "Content-Type: application/vnd.drawbridge.entry.v1+json" -d '{"digest":{}}' "localhost:8080/${1}/_tag/${2}"

# Upload package
curl --fail-with-body -sX PUT -H "Content-Type: application/vnd.drawbridge.directory.v1+json" -d '{"Enarx.toml":{"digest":{}},"main.wasm":{"digest":{}}}' "localhost:8080/${1}/_tag/${2}/tree"

# Upload workload
curl --fail-with-body -sX PUT -H "Content-Type: application/wasm" -T "${3}" "localhost:8080/${1}/_tag/${2}/tree/main.wasm"

# Upload config
curl --fail-with-body -sX PUT -H "Content-Type: application/toml" -T "${4}" "localhost:8080/${1}/_tag/${2}/tree/Enarx.toml"
```

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -64,42 +77,121 @@ fn main() -> Result<()> {
...
+                .deref()
+                .iter()
+                .find(|w| w.with_backend(backend))
+                .ok_or_else(|| anyhow!("no supported exec found"))
+                .map(|b| b.exec())?;
+
+            let (exec_sock, mut host_sock) =
+                UnixStream::pair().context("failed to create a Unix socket pair")?;
+            assert_eq!(exec_sock.as_raw_fd(), 3);
+            assert_eq!(host_sock.as_raw_fd(), 4);
```
Comment (bstrie) [inline src/main.rs:98]: Magic number comments.

File: src/main.rs [hunk 2]
```diff
@@ -64,42 +77,121 @@ fn main() -> Result<()> {
...
+                "file" => {
+                    let path = package.to_file_path().map_err(|()| {
+                        anyhow!("failed to parse file path from URL `{}`", package)
+                    })?;
+                    match fs::metadata(&path).with_context(|| {
+                        format!(
+                            "failed to get information about `{}`",
+                            path.to_string_lossy(),
+                        )
+                    })? {
```
Comment (bstrie) [inline src/main.rs:110]: I find it difficult to read the match with such a long expression here, can we split this out into its own binding? This match is already not really using patterns, so then we can just have `if md.is_file() { ... } else if md.is_dir() { ... } else {`

File: src/main.rs [hunk 3]
```diff
@@ -64,42 +77,121 @@ fn main() -> Result<()> {
...
+            exec_io
+                .join()
+                .expect("failed to join exec-wasmtime I/O thread")?;
+            drop(exec_sock);
+            if let Some(wasm) = wasm {
+                drop(wasm)
+            }
+            if let Some(conf) = conf {
+                drop(conf)
+            }
```
Comment (bstrie) [inline src/main.rs:194]: Are these if-lets necessary? Dropping a None should be a no-op.

File: Cargo.toml [hunk 4]
```diff
@@ -31,16 +31,18 @@ bitflags = { version = "1.2", default-features = false }
 clap = { version = "3.1", features = ["env", "derive", "std"], default-features = false }
 colorful = { version = "0.2", default-features = false }
 dirs = { version = "4.0", default-features = false }
+enarx-exec-wasmtime = { version = "0.4.0", path = "crates/exec-wasmtime", default-features = false }
```
[the remark this answers is not in the document]
Comment (rvolosatovs) [reply Cargo.toml:34]: Sure, please file an issue and let's address in a follow-up

File: crates/exec-wasmtime/src/lib.rs [hunk 5]
```diff
@@ -10,73 +10,68 @@
...
+        wasm: RawFd,
+        /// Optional open config file descriptor
+        conf: Option<RawFd>,
+    },
+}
 
 /// The Arguments
-#[derive(Parser, Debug)]
+// NOTE: Order of fields matters for this struct and `Package` seem to be required to be the last
+// field, otherwise `toml` serialization fails with `values must be emitted before tables`
```
[the remark this answers is not in the document]
Comment (rvolosatovs) [reply crates/exec-wasmtime/src/lib.rs:42]: I would guess that implementation does some kind of reflection over the fields here and it would not matter, but I'll check.

File: crates/exec-wasmtime/src/lib.rs [hunk 6]
```diff
@@ -10,73 +10,68 @@
...
     let requested = configured.next()?;
     let attested = requested.next()?;
     let acquired = attested.next()?;
     let compiled = acquired.next()?;
     let connected = compiled.next()?;
     let completed = connected.next()?;
-
     drop(completed);
+
+    forget(host);
```
[the remark this answers is not in the document]
Comment (rvolosatovs) [reply crates/exec-wasmtime/src/lib.rs:74]: The FD is opened by the host and the host should handle closing it as well.
If we were to forget in the host instead, then if `exec-wasmtime` panics, for example, we could be leaking FDs.

File: crates/exec-wasmtime/src/loader/attested.rs [hunk 7]
```diff
@@ -1,25 +1,116 @@
...
+                match res.content_type() {
+                    WASM_MEDIA_TYPE => response_into_wasm(res).map(|webasm| (webasm, None))?,
+                    DRAWBRIDGE_DIRECTORY_MEDIA_TYPE => get_drawbridge_directory(url.as_str())
+                        .map(|(webasm, config)| (webasm, Some(config)))?,
+                    t => bail!("unsupported content type: {}", t),
+                }
+            }
+            Package::Local { wasm, conf } => {
+                let mut webasm = Vec::new();
+                unsafe { std::fs::File::from_raw_fd(wasm) }
```
[the remark this answers is not in the document]
Comment (rvolosatovs) [reply crates/exec-wasmtime/src/loader/attested.rs:89]: When we talk about safety here, what do we mean? Is it:
1. the fact that we're not going to segfault here that we're trying to assert
or,
2. are we talking about integrity of data and the whether or not we trust that this FD is indeed the FD containing WASM that the user wanted us to execute, i.e. no one has tampered with the data and this is indeed what the user wanted to execute.

Well, there's no way to programmatically validate that this is a valid FD AFAIK (@haraldh any ideas ?).
We assume that this executable was executed by the host binary, we assume that our implementation is correct, and, therefore, the FD opened by the host binary (via a safe operation) was passed to this executable and this operation *should* be safe. However, due to (2.), there's no guarantee that the opened FD by the host was not tampered with on the way.

We got this FD from the host and by definition we should not be trusting what host tells us, but we are trying to run a workload from a local FS here, which is insecure already and requires trust in the host anyway.
So basically, the only way this can be safe is if we trust the host and we don't, but I also don't think that this should really warrant marking `next()` unsafe, since that'd mean that essentially everything `exec-wasmtime` does is unsafe.

I'll keep the unsafe annotations as-is and add a SAFETY comment here as outlined above, but again, technically this whole thing is unsafe.

File: crates/exec-wasmtime/src/loader/attested.rs [hunk 8]
```diff
@@ -1,25 +1,116 @@
...
+                    .read_to_end(&mut webasm)
+                    .context("failed to read WASM module")?;
 
+                let config = if let Some(conf) = conf {
+                    let mut config = String::new();
+                    unsafe { std::fs::File::from_raw_fd(conf) }
```
[the remark this answers is not in the document]
Comment (rvolosatovs) [reply crates/exec-wasmtime/src/loader/attested.rs:95]: See above

File: src/main.rs [hunk 9]
```diff
@@ -64,42 +77,121 @@ fn main() -> Result<()> {
...
+        cli::Command::Run(run::Options {
+            backend,
+            package,
+            steward,
+            #[cfg(feature = "gdb")]
+            gdblisten,
+        }) => {
+            let backend = backend.pick()?;
+            let exec = EXECS
+                .deref()
```
[the remark this answers is not in the document]
Comment (rvolosatovs) [reply src/main.rs:89]: That was existing implementation, which I did not change. We can refactor in a follow-up.

File: tests/enarx_wasm/mod.rs [hunk 10]
```diff
@@ -260,13 +251,20 @@ fn check_tcp() {
...
+    fs::copy(
+        env!("CARGO_BIN_FILE_ENARX_WASM_TESTS_check_tcp"),
+        pkg.path().join("main.wasm"),
+    )
+    .expect("failed to copy WASM module");
+
+    fs::write(
+        pkg.path().join("Enarx.toml"),
+        CFG.replace("@@LPORT@@", &cport.to_string())
+            .replace("@@CPORT@@", &lport.to_string()),
```
[the remark this answers is not in the document]
Comment (rvolosatovs) [reply tests/enarx_wasm/mod.rs:265]: Nope, again, no changes here from the existing implementation.
The confusion arises here from the fact that we are describing two parts of a connection, i.e. we spawn a thread to listen on `lport`, but we need our workload to connect on that `lport` (which is `CPORT` from the perpective of the workload/config)
I agree that the variable naming is not clear here, but again, something for a follow-up to keep the diff small, I'll file an issue.

File: crates/exec-wasmtime/src/loader/attested.rs [hunk 11]
```diff
@@ -1,25 +1,116 @@
...
+        .read_to_end(&mut wasm)
+        .context("failed to read WASM module contents")?;
+    // TODO: Verify Content-Digest
+    Ok(wasm)
+}
+
+fn get_drawbridge_directory(url: impl AsRef<str>) -> Result<(Vec<u8>, String)> {
+    let url = url.as_ref().trim_end_matches('/');
+    // TODO: refactor into `get_typed<WASM_MEDIA_TYPE>` once `&str` is supported as
+    // const generic argument
```
[the remark this answers is not in the document]
Comment (rvolosatovs) [reply crates/exec-wasmtime/src/loader/attested.rs:40]: What value would it bring?
It does not really help the particular issue, since enum variants cannot be used as generic arguments and we need strings anyway.
I briefly experimented with a trait containing an associated type, but that was too awkward, string is the simplest approach.

File: Cargo.toml (hunk 4 above)
[the remark this answers is not in the document]
Comment (haraldh) [reply Cargo.toml:34]: Here is the follow-up:
https://github.com/enarx/enarx/pull/1847

File: crates/exec-wasmtime/src/lib.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (rvolosatovs) [reply crates/exec-wasmtime/src/lib.rs:74]: yes, done locally already

File: src/main.rs [hunk 12]
```diff
@@ -37,6 +49,78 @@ struct Options {
...
+            .shutdown(Shutdown::Read)
+            .context("failed to shutdown read half of host's socket")
+    });
+
+    let exit_code = keep_exec(backend, backend.shim(), exec, gdblisten)?;
+    exec_io
+        .join()
+        .expect("failed to join exec-wasmtime I/O thread")?;
+    drop(exec_sock);
+    files.for_each(drop);
```
Comment (haraldh) [inline src/main.rs:120]: This should not be necessary.

Comment (haraldh) [inline src/main.rs:120]:
```suggestion
```

File: tests/wasm/mod.rs [hunk 13]
```diff
@@ -190,102 +253,124 @@ fn no_export() {
...
-    let bin = env!("CARGO_BIN_FILE_ENARX_WASM_TESTS_echo");
-    run_wasm_test(bin, None, 0, input, input, None);
+    run_wasm_test(
+        env!("CARGO_BIN_FILE_ENARX_WASM_TESTS_echo"),
+        None,
+        0,
+        input,
+        input,
+        None,
+    );
```
Comment (haraldh) [inline tests/wasm/mod.rs:269]: Back then I introduced `bin` to remove the line breaks

Comment (haraldh) [reply tests/wasm/mod.rs:269]: Why, oh why did you add them back everywhere?

File: tests/wasm/mod.rs [hunk 14]
```diff
@@ -190,102 +253,124 @@ fn no_export() {
...
+prot = "tcp"
+port = @@LPORT@@
+name = "LISTEN"
+
+[[files]]
+kind = "connect"
+prot = "tcp"
+host = "127.0.0.1"
+port = @@CPORT@@
+name = "CONNECT""#;
```
Comment (haraldh) [inline tests/wasm/mod.rs:341]: What's your problem with the old form? It was more readable than this...

**Review by haraldh — CHANGES_REQUESTED** [review]

File: tests/wasm/mod.rs (hunk 13 above)
Comment (rvolosatovs) [reply tests/wasm/mod.rs]: `bin` is a redundant variable, which is only used once and duplicated for N test cases we have.
1. When reading the code, one has to check what `bin` is - look through the function and check for usages. That is not required when there is no variable.
2. It's faster to duplicate and adapt a single invocation, rather than a variable binding and an invocation (note that it has to be done for N test cases)

I disagree, but I will add a `const` here

**Review by haraldh — APPROVED** [review]
